### PR TITLE
Feat: Threaded MutationsBatcher

### PR DESCRIPTION
**Pull request by Mariatta**
- Batch mutations in a thread to allow concurrent batching
- Flush the batch every second
- Flow control
- Batcher can now be used as a contextmanager

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigtable/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:62]: We should also block when the total inflight size is > 100MB. I think we'll need another variable FLUSH_SIZE to track when the current batch is full, and track the total size separately.

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:256]: how are errors handled? the java client throw errors on the failed elements onClose(). Will the batcher fail silently if some elements failed to apply?

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:37]: as a user, how can i reconfigure max bytes and flush count? (This can come in a later pr)

Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:37]: Sure, I will update the numbers to match the ones set for Java.

Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:37]: These values gets passed from the `MutationsBatcher` constructor.

For example, they can do:
``` 
batcher = MutationsBatcher(max_row_bytes=5*1024, flush_count=50)
```

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:256]: It uses the existing retry mechanism in `table.mutate_rows`.

https://github.com/googleapis/python-bigtable/blob/main/google/cloud/bigtable/table.py#L680

Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:256]: So `table.mutate_rows` returns the response statuses to indicate errors.
We can capture the status codes and raise error.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:57]: This is the max number of mutations for a single row. (100,000) In this case I think it should raise an error since we shouldn't split the mutations for one row?

File: google/cloud/bigtable/batcher.py
Comment (igorbernstein2) [inline google/cloud/bigtable/batcher.py:25]: Please add comments what these constants control

File: google/cloud/bigtable/batcher.py
Comment (igorbernstein2) [inline google/cloud/bigtable/batcher.py:55]: when would block = False?

File: google/cloud/bigtable/batcher.py
Comment (igorbernstein2) [inline google/cloud/bigtable/batcher.py:148]: Please remove the TODO from the doc
Also please document the threadsafety behavior. From what I see in the code, a Batcher can't be shared by multiple enduser  threads?

File: google/cloud/bigtable/batcher.py
Comment (igorbernstein2) [inline google/cloud/bigtable/batcher.py:279]: will this throw the original exception or will it wrap it?
If its the original exception, then the stacktrace will be rooted in the executor. If its wrapped then you are leaking implementation details.

In java we kept the original exception but added a synthetic stacktrace of the caller to help callers diagnose where they called the failed RPC.

HBase does something even hackier and modifies the stacktrace and inserts the callers stack at the top.

I dont know which approach idiomatic to python, but we should be intentional here

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]
It seems like we don't really have any flow control here. Do you plan on adding it in a follow up PR?

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:279]: This will re-raise the original exception. I will see how to do that in Python.
Can you share link on how it was done in Java?

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [inline google/cloud/bigtable/batcher.py:303]: I noticed Status codes can be converted into exceptions using [google.api_core.exceptions.from_grpc_status](https://github.com/googleapis/python-api-core/blob/b6eb6dee2762b602d553ace510808afd67af461d/google/api_core/exceptions.py#L441). Maybe it would be better to raise those directly, to give the full context, rather than using the raw status codes?

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:279]: In this case it the exception will have the stack trace all the way to the original spot where it was raised (line 303 below). The exception itself will have a list of individual error codes so user can still go through it to find out what failed.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:160]: I've removed the TODO. Correct the same batcher cannot be shared by multiple enduser threads.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:333]: An error is now raised if any of the mutation errored out.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:38]: Yes I've thought of the exception group and as you know it's only in Python 3.11. Your suggestion sounds great, I'll make the changes.

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:102]: Why do we need to check if the queue is full? In the batcher code, we flush when the Batcher is full:
```
if self._rows.full():
    self.flush_async()
```
I'm not sure if the behavior is still correct? 

I think full() should only check for 2 things: 
- Number of elements reached the batch element threshold
- Number of bytes reached the batch bytes threshold

The queue is only be used to block user from adding more elements when it's full. We don't trigger another flush when the queue is full.

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:268]: would this also remove the row from the queue? In this case, we can't block on adding more elements to the queue, or when we block it doesn't necessarily mean that we've reached the max outstanding bytes or elements?

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:286]: I think we may want to only raise the errors when we close the batcher

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [inline google/cloud/bigtable/batcher.py:369]: is this meant to be public?

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [inline google/cloud/bigtable/batcher.py:417]: Maybe also unregister the `atexit` callback registered in `__init__`?

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [inline google/cloud/bigtable/batcher.py:338]: I'm a bit confused by `BatchInfo`. Doesn't it duplicate the `row_count`, `mutations_count` and `mutations_size` variables?

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [inline google/cloud/bigtable/batcher.py:380]: Having both `flush()` and `flush_rows()` seems a little confusing. Especially since this doesn't seem to be "flushing" from the cache in any way. I could see people calling this with no arguments thinking it is the main flush function

Maybe this should be called `mutate_rows`? Or just made into an internal helper function?

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [inline google/cloud/bigtable/batcher.py:388]: this would raise a `TypeError` if `flush_rows` is called without any arguments. Is that the intention?

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:353]: I dont think we need to check flow control limits here since you're calling `self.flow_control.control_flow(batch_info)` on line 357.

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:355]: why not call _rows.isfull()?

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:369]: Changed to private.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:388]: It was a mistake that the `rows_to_flush` was optional. I've changed it to required.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:417]: Yes, added.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:380]: Agree this is confusing. There is already a public `mutate_rows` from before, and it has different behavior than this one. I'm changing it to private since user isn't expected to call it manually.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:355]: In here, we're popping an elements from `self._rows` (line 341), taking items out from the queue, perform mutations in a batch, and waiting for response.
We can't call `_rows.full()` because it is being emptied at this point, so it doesn't become full.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:353]: We still need to check it here. At this point the flow control has not been activated yet. We're checking if we will need to activate the flow control after this condition is met. Without this check it will not actually activate the flow control.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:338]: It's similar info, but for different "buckets".

There are two "queues":

`self._rows`: for storing the rows we want to mutate
`batch_info`: for storing info about the rows that are being mutated, and we're waiting for the result/response from backend. This gets passed to the the batch_completed_callback. That's where we can release the flow control.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:322]: Yes rows gets popped out of the queue. I have added flow control mechanism after your comment about this. So this is outdated.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:149]: I was using this for debugging. I will remove it in the end before merging.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:74]: Since queue.put (line 87) is blocking and waiting until it becomes free again, I'm removing the errors.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:80]: queue.put() is already blocking and waiting until it has space available. So I'm removing this error.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:116]: And these numbers are:
MAX_OUTSTANDING_ELEMENTS  = 100000
MAX_MUTATION_SIZE = 20 MB 
correct?

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py]: This has been addressed.

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:47]: nit:
```suggestion
    def __init__(self, max_mutation_bytes=MAX_MUTATION_SIZE, flush_count=FLUSH_COUNT):
```

File: google/cloud/bigtable/batcher.py
Comment (mutianf) [inline google/cloud/bigtable/batcher.py:97]: defaults should be:
```suggestion
        self, max_mutations=MAX_OUTSTANDING_ELEMENTS, max_row_bytes=MAX_OUTSTANDING_BYTES
```

**Review by mutianf — APPROVED** [review]
LGTM with some nits!

**Review by daniel-sanche — APPROVED** [review]

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py]: Deleted

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:95]: Sure, I will add that in the next commit.

File: google/cloud/bigtable/batcher.py
Comment (daniel-sanche) [reply google/cloud/bigtable/batcher.py:112]: Since I also was looking at something similar for the v3 branch: Python semaphores only allow incrementing by 1 at a time, but we'd need to increment it by the thousands to track bytes.

My current solution was to use a [condition](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Condition), with a check against the flow limits as a predicate. But I think the event can be set up in a similar way

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py]: I adjusted the logic, moving the `wait` to the `flush_async` function. If the batch is causing the event to be blocked, then it will be sent through, but the subsequent thread will be blocked and waited.

File: google/cloud/bigtable/batcher.py
Comment (Mariatta) [reply google/cloud/bigtable/batcher.py:290]: This function is executed per second, and when the batch gets closed, when we're supposed to send all remaining batches sequentially.

**Review by igorbernstein2 — APPROVED** [review]
LGTM, but please have Daniel take a look at the api and confirm it works for him in the async client

**Comment by daniel-sanche** [comment]
Some of the API will need to change for the async client, due to different model classes and asyncio patterns, but the general shape of the solution should be mostly consistent. LGTM

**Pull request merged**
